From e3f937751fcfa16b809066500e09ec930332877f Mon Sep 17 00:00:00 2001
From: "[Partner] Sergey Nizovtsev" <sn@tempesta-tech.com>
Date: Sat, 25 Jun 2022 09:01:06 +0300
Subject: [PATCH] GCORE: bpf: allow dynamic kfuncs in XDP programs

Tell verifier to search allowed kfunc BTF IDs against dynamic set.

This patch allow modules to register their own helpers for XDP programs:

    BTF_SET_START(mod_kfunc_ids)
    BTF_ID(func, helper1)
    BTF_ID(func, helper2)
    BTF_SET_END(mod_kfunc_ids)

    register_kfunc_btf_id_set(&prog_test_kfunc_list, &mod_kfunc_btf_set);
    unregister_kfunc_btf_id_set(&prog_test_kfunc_list, &mod_kfunc_btf_set);

We hack prog_test_kfunc_list here as it is the only BTF set declared in
vmlinux. Linux-5.18 introduces a more generic way to register kfuncs so
this patch should be removed.

Signed-off-by: Sergey Nizovtsev <sn@tempesta-tech.com>
---
 net/core/filter.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/net/core/filter.c b/net/core/filter.c
index f8fbb5fa74f35..4c3a0bac1879f 100644
--- a/net/core/filter.c
+++ b/net/core/filter.c
@@ -10094,6 +10094,7 @@ const struct bpf_verifier_ops xdp_verifier_ops = {
 	.is_valid_access	= xdp_is_valid_access,
 	.convert_ctx_access	= xdp_convert_ctx_access,
 	.gen_prologue		= bpf_noop_prologue,
+	.check_kfunc_call	= bpf_prog_test_check_kfunc_call,
 };
 
 const struct bpf_prog_ops xdp_prog_ops = {
-- 
2.38.0

